### DISCUSSION TITLE
Official Support for Valhalla Build Artifacts for iOS and Android

### DISCUSSION THREAD

**Discussion by Archdoog** [valhalla/valhalla#4509] [Ideas]
There's been substantial discussion about producing valhalla builds for iOS and Android mobile applications over the years. This is primarily for the purpose of generating routes while offline from a pre-processed valhalla tile data set.

This issue aims to understand the subtasks and requirements for getting this code directly contributed to the valhalla repo. This will ideally include create valhalla artifacts directly out of CI for iOS and Android. Additional requirements like an Objective C++ or JNI layer are likely to be up to individual developers downstream, but some basic readme docs may be valuable.

## Goals

For each of these, I'm happy to create an individual issue ticket to break up the work and focus any discussion if there's agreement on the desired outcome of building PR(s) directly for this repo.

- [ ] ~~Determine which features should be enabled to support the desired requirements (e.g. simply creating a valhalla/tyr/actor instance.). Do we need additional flags in the cmake to toggle or bundle things appropriately.~~
- [ ] ~~Determine how to bundle standalone dependencies to support the desired output (boost headers, protobuf and potentially others like curl). Can we build in CMakeLists functionality to directly fetch and include these from a source like the third party repos when they don't exist on a platform?~~
- [ ] ~~Add `SHARED` flag handling and create a shared library artifact for Android (e.g. `libvalhalla.so`). Alternatively, maybe this needs to be built in a wrapper project?~~
- [ ] ~~Generate static libraries for iOS and merge into xcframework (e.g. `libvahalla.xcframework`).~~
- [ ] ~~Create Android integration readme file. This should explain the basics of to put the output into an android library or project (e.g. `libvalhalla.so`)~~
- [ ] ~~Create an iOS integration readme file. Same as above with the `libvalhalla.xcframework`.~~
- [ ] ~~Add ios and android build scripts to CI pipeline and attach artifacts?~~

> [!IMPORTANT]  
> See the latest comment with an update & solution https://github.com/valhalla/valhalla/discussions/4509#discussioncomment-11023152

## How To

With the goals above in mind, there needs to be some discussion about if and how this can be contributed back to the main valhalla repo. The goal here is not to add the objective c++ layer or JNI to the repo, but to provide the build scripts/instructions to build these platform ouputs. Most of this might be focused on enabling the `CMakeList.txt` to actually toggle and require validate the flags for each platform output. In the simplest case, this may be just producing the existing documentation.

Having done some parallel work (building a working swift package that directly imports valid c++ similar to how firebase-ios-sdk works), I think it's ideal from a long term maintenance standpoint to get this contributed back to the main repo. The scope of valhalla and it's dependencies makes it easy to diverge over time with external projects.

## History

The following includes a history of issues discussing this topic:

https://github.com/valhalla/valhalla/issues?q=is%3Aissue+iOS+OR+Android+

Let me know what everyone thinks.

**Comment by nilsnolde** [comment]
sorry for the wait.. let me give my opinion on this:

first of all, it'd be very valuable for everyone to have a guide/example repo how to build mobile apps with native libvalhalla support, so definitely +1.

I'm currently overhauling our CMakeLists.txt with the overall goal to make libvalhalla available again to package managers, most notably for your use case including `vcpkg` (having support for android/ios toolchains). most importantly, we'll want cmake to package valhalla properly, resolving static and shared libraries. some of that is already possible (e.g. building a shared lib via `-DBUILD_SHARED_LIBS` or even building each module into a static library via `-DENABLE_STATIC_LIBRARY_MODULES`). but currently we don't have a way to package _both_, which might be useless for your use case. still, you could run two builds to get the artifacts and copy them around for now. I don't think we should be the one making any artifacts available (other than docker) for specific platforms, we should push that to package managers like vcpkg.

I think the possibly best way to go would be:
- prioritize to make libvalhalla available via vcpkg so it can easily be installed on any supported platform; not much should be missing here as long as we can live with just having only the static library libvalhalla.a available, if not, we'll have to solve installing/packaging both at the same time. good news is that vcpkg has TONS of existing ports of c++ applications, probably even qgis very soon
- then it'd be nice to have separate repos in the valhalla org for both ios & android installing everything via vcpkg, providing an example app for both including the build gymnastics and docs

I'd (for now) advocate against focusing on making valhalla compatible with a pure CMake build via add_subdirectory or Fetch* API and rather focus on general package manager compatibility. we'd also prioritize that projects are able to link to libvalhalla via pkg-config vs having our own valhalla-config.cmake installables.

finally, I should mention that I have zero experience with building for mobile platforms:)

**Comment by Archdoog** [comment]
Hello Nils, no problem! You actually gave me just enough time to finish my android poc and to successfully generate a route from a tar on android: https://github.com/Rallista/ValhallaAndroidPOC (see the `src/` folder for the valhalla c++ stuff, the `valhalla/` folder is the android wrapper). It's likely following the learnings from that will make a similar iOS build relatively simple as well. Unfortunately, the final solution was exactly what you advocated against 😄. Either way, this should be useful to help inform a path forward.

1. Protobuf really was the main challenge with mobile builds. I tried 3 or 4 cmake arrangements and the only one that didn't end in some roadblock was the Fetch API. 
2. Shared vs static. Android requires a shared library unless you want to manage a bunch of hacky loaders. Static libraries can only be imported from absolute file paths for some reason, so you basically have to fully handle and understand all the nuances of various android manufacturer's filesystems. iOS, being a c based platform on the other hand can easily just bundle a bunch of static libs and call right in if they're linked to the Xcode project.
3. Lastly for build gymnastics, this POC was informative. I think there are a handful of paths forward. It may make the best sense to just improve the cmake files to more easily allow pulling in protobuf from source (e.g. the Fetch api) and allow some flags for android/iOS to enforce the expected output (static & shared). Also, finding a way to explicitly differentiate protoc for your host machine, libprotobuf for your build target and validate version checks. Aside from that, I actually found it quite helpful on android to treat valhalla as a submodule in an android library module with the `src/wrapper` in my repo.

My next steps are to reproduce this POC as an actual android library to see how much more labor goes into integrating the cmake build into gradle to automatically handle it locally. Since android gradle supports this, it would just let the developer add the android dependency and then the android build system would generate whatever shared libraries it needs for supported architectures by calling into cmake.

Last note for now, in the spirit of finally solving this for the Valhalla rep. I'm more than happy to contribute back whatever you see as useful.  

I'll probably have more to say this weekend/next week. Just wanted to give you an update since things went well.

**Reply by nilsnolde** [reply]
Nice, congrats:) I'd love to try this myself some day. For now I'd focus myself on the packaging stuff. As alluded to previously, I'd like to be as little as possible dependent on CMake API (I also never tried the Fetch API ever). And instead rely on decades-old software like package managers and pkg-config which won't go anywhere and don't have these hugely complicated/opaque things as CMake. It's good to have it manage the interplay of available dependencies on a system, but IMO those should come from the system itself (or vcpkg for inter-platform convenience). Fetch can maximally be an additional way to manage the project IMO. Though vcpkg is of course also external, but it's basically "just" a CMake wrapper to do the installation/packaging, which has to be done by _some_ software (used to be autotools I guess, vcpkg/CMake do make that easier syntactically though). 

Anyways, that's just my thoughts on the rather long-term approach and your POC is a great starting point!

**Reply by farfromrefug** [reply]
Just for reference Carto Mobile SDK is successfully building and using valhalla. https://github.com/CartoDB/mobile-sdk
I know we use a fork of valhalla to allow us to read valhalla tiles from a compressed sqlite format. I can give more info if wanted

**Comment by Archdoog** [comment]
Finally wrapped up all the tooling to create an officially published iOS and android library for Valhalla! It still uses a wrapper [architecture](https://github.com/Rallista/valhalla-mobile/blob/main/docs/src/architecture.md) to conveniently encapsulate Valhalla functionality that should be exposed to mobile including JNI for Android. Currently, only route is available, but happy to accept contributions for whatever else people need now that all the plumbing, CI and release publishing is handled. 

### What's There

- The core project https://github.com/Rallista/valhalla-mobile/ which is published as a Swift Package and an android library on Maven Central. See the [README.md](https://github.com/Rallista/valhalla-mobile/blob/main/README.md) for a good overview.
- https://github.com/Rallista/valhalla-openapi-models-swift - This provides all the valhalla API and Config JSON models (by no means 100% correct/finished). Giving the swift Valhalla-mobile library typed inputs and outputs from libvalhalla through the wrapper.
- https://github.com/Rallista/valhalla-openapi-models-kotlin - Same as the swift package, but for kotlin.
- https://github.com/stadiamaps/osrm-openapi - Same as the valhalla packages, but for OSRM flavored since most navigation SDKs lean that way.

### The iOS and Android testing that proves it

- [TestValhallaWithFolder.swift](https://github.com/Rallista/valhalla-mobile/blob/main/apple/Tests/ValhallaTests/TestValhallaWithFolder.swift) - this validates the full end to end with a pre-downloaded tar folder on iOS.
- [TestValhallaWithTar.swift](https://github.com/Rallista/valhalla-mobile/blob/main/apple/Tests/ValhallaTests/TestValhallaWithTar.swift) - Same as above with a tar.
- Similar but a bit different on Android: https://github.com/Rallista/valhalla-mobile/tree/main/android/valhalla/src/androidTest/java/com/valhalla/valhalla.

### Next steps

I'll probably have a few small PRs for this repo that simplify the process and to add a link in this repo's README.md if there's interest.

I'm sure the valhalla open api files could provide more value outside of this limited use case. Happy to help if that's in scope for this organization to post and maintain.

Aside from that, this seems to pretty reproducible and easy to use at this point. Just need to build some docs, incrementally improve everything, increase its scope of features and optimize some stuff.

Thanks for the discussion on the topic. Was definitely helpful in getting it to this point!

**Reply by ianthetechie** [reply]
Man, it's so exciting to see this all come together after so many years of fragmented efforts 😂 Thanks for all your work on this, and I'm happy to have had a small part in it. Now if only we could figure out a better approach to #2960 so that we didn't have to maintain all these separate OpenAPI models haha.

**Reply by kevinkreiser** [reply]
excellent work! long term id still like to have, as ive said for the past many years hahha, android and ios targets in the main project each with a small top level smoke test, like we do for the python bindings, to check that every api does what its broadly supposed to do. of course i never have any free time so i cant say when as the build system fun always takes eons!

**Reply by Archdoog** [reply]
That's a reasonable goal. 

The core cmake stuff in here https://github.com/Rallista/valhalla-mobile/blob/main/src/CMakeLists.txt isn't terrible and could probably be translated into the main valhalla CMakeList.txt without too much work. Most of it is just ensuring the third party libraries actually exist and can be linked. Plus some funky platform specific stuff like the tzdata.tar file on iOS. Putting this in the core Valhalla library could also encourage some optimizations that are beneficial to mobile devices (e.g. disabling the whole build tiles portion of the library).

The bigger issue with iOS (lesser) and Android (moreso) is the binding layer(s), library packaging and publishing. With Swift C++ things are pretty concise. But properly packaging an iOS `.xcframework` linked to a swift package version is a bit of CI/CD pain. On Android with Java/Kotlin, this is going to be a mess of JNI code plus the joys of publishing to MavenCentral. On both platforms, to get a basic primitive layer wouldn't be too bad, but I doubt the full typed support in [valhalla-mobile](https://github.com/Rallista/valhalla-mobile) would make sense in this library.

All of this is doable, it's just hard to decide where to draw the line and the variance in level of effort can become quite substantial. Especially when platforms like flutter, etc, may also draw appeal.

**Reply by kevinkreiser** [reply]
yeah to be clear i dont want to do any of the packaging stuff :smile: the minimum and probably maximum of what i'd like here would be to build the bindings jni/swift (just like we do for python) and then have a small test in java/swift that just hits each main API (the ones we have in `actor.h`) to prove it works. for both bindings the input and output would be json or pbf strings to keep it simple. side note, ive done plenty of the jni stuff before and its not so bad but i have no experience with swift and simultaneously i hate the whole apple ecosystem haha but i feel we should (reluctantly) support it because of its prevalence

**Comment by farfromrefug** [comment]
@Archdoog great job. There is already a working port of Valhalla to mobile through https://github.com/CartoDB/mobile-sdk. It is not maintained anymore. But i have started to officially support a fork which i will rename in the months to come.
The approach is a bit different in the sense that it does not provide direct binding usable as a separate module (though i plan on doing it). It is part of the SDK.
The reason why i am presenting this here is more because of the approach that was taken to handle the valhalla data. The SDK does not use tar files directly. Instead we generate a mbtiles from valhalla tiles. We had to modify valhalla a bit to be able to read from that mbtiles (the diff is quite minor). The really good thing about this is that that mbtiles is compressed (gzip tiles) which reduces the size by a LOT. For europe:
* valhalla_tiles.tar: 31.4Gb
* valhalla_tiles.mbtiles: 14 Gb

And the computation time is not really that much slower (though i dont have time comparaison).

Just something to think about if you go mobile!
Here is the script we use to generate the mbtiles: https://github.com/Akylas/alpimaps_data_generator/blob/master/scripts/build_valhalla_package.py

**Reply by ianthetechie** [reply]
FWIW, Valhalla also includes support for LZ4 and I think GZIP compression on individual tiles in a directory. IMO for mobile this is simpler than MBTiles since it doesn't involve a database layer on top and is easier to just swap individual tiles in and out. No transformation required.

**Reply by kevinkreiser** [reply]
@ianthetechie is correct, it also allows one to easily download individual tiles on an as needed (as an algorithm encounters disk cache misses for specific tiles) basis if you set up a tileserver that serves gzipped tiles

**Reply by farfromrefug** [reply]
@ianthetechie @kevinkreiser did not know that. Does that mean you can generate valhalla_tiles.tar compressed ?
As for the downloading of individual tiles or server this does not apply to mobile. The only thing you want is small disk usage and memory footprint. It is almost always for offline usage

**Reply by kevinkreiser** [reply]
no its not a compressed tar, its compressed individual tiles. the tile server is exactly the use-case mobile was meant for. basically it looks like this:

1. you build a global tileset, dont tar it
2. you gzip/lz4 each tile in the set
3. you start up a tileserver (it just serves the tile directory with the individually compressed tiles from step 2)
4. your mobile client has no tiles on it but it does have, `tile_dir` configured (thats the disk cache) and `tile_url` configured, thats the url to your tile server

in this way when you ask for a route from teh library valhalla will fetch just the data it needs to compute the route on the fly and it will cache it in the `tile_dir` on the mobile device so that subsequent requests will be able to rely on this disk cache before hitting the network.

there are other somewhat complex considerations like when to clear the disk cache and which version of tiles exists on the tile server vs on the client (different builds of the data will not connect to each other properly so you cant mix the tiles) but these problems are solvable once you get serious about running the router in this configuration.

**Reply by farfromrefug** [reply]
@kevinkreiser Ok but that is for when you use a remote valhalla server from mobile. I thought this all discussion was about having offline valhalla computing on mobile (which is what we do in the SDK). IN this case we want tiles compressed inside a tar file (much faster to read a tar file than having thousands of file on disk, especially knowing how bad SAF is).
Yes in the SDK we have another build step in creating an mbtiles. But thanks to the sqlite format we get much faster read than while using disk files (on android even more thanks to SAF again..)

**Reply by dnesbitt61** [reply]
If offline tiles are stored in ./mbtiles, how would you add tiles and delete tiles from the dataset? Say I have Valhalla tiles in a .mbtiles dataset for several US states loaded. Now I need to add tiles for a new US state/region. Or to remove tiles for a state/region I no longer need? Are these operations faster than having a set (perhaps large number) of compressed Valhalla tiles on device?

**Reply by ianthetechie** [reply]
You aren't wrong. I just don't think it matters 😅 A mobile device isn't going to be computing routes at continent scale at a rate that it matters. Yea a single tar is one file handle and this actually does make a difference when mmaping on a server with a crap load of RAM, but it seems like needless complexity on mobile. I'd say the same of an MBTiles archive. This is actually more interesting as a server side technology for doing updates... maybe.... Kevin also points out that this has challenges too...

**Reply by farfromrefug** [reply]
@dnesbitt61 we dont add tiles. we add mbtiles per region and then valhalla reads them all as a whole. WE already have a Package service which can be used as a server provider for mbtiles.

@ianthetechie well then what s the point of building valhalla for mobile if you depend on a valhalla remote server to compute the routes?
And no it wont be computing at continent scale (which my server wont either in a short enough time). But tbh continent scale routing is not the primary use case. I dont think a lot of people go beyond the 500km llimit which is set by default in the valhalla config. And 500km is working quite fine on mobile. I even often compute 4 routes alternatives simultaneously to get alternatives.
I use valhalla offline routing daily on my 2017 phone which is now really old.

**Reply by dnesbitt61** [reply]
So if you have multiple regions on device, Valhalla has to be able to read from the correct one (or read from all)? For example, if I have a route that spans Maryland, Pennsylvania, and Delaware and have downloaded tiles for these regions, it seems more straightforward to simply add all tiles from the multiple downloaded regions to a single set of files (compressed) on mobile storage. Then Valhalla simply reads the set of tiles it needs (uncompressing as it first accesses them) and retains them in the memory cache. I can't imagine querying multiple mbtiles files to find the correct one to read is faster than simply reading from file storage.

**Reply by farfromrefug** [reply]
@kevinkreiser It reads mutilple mbtilws to query all tiles then Valhalla works just ad it would with a tar file. 
Now about your speed remark. It will be much faster with few mbtiles instead of thousands of disk tiles files especially on latest Android versions (>=10). Android now uses SAF which makes files operations crazy slow ( searck liké for " Android sad slow" to see what in talking about).

**Reply by ianthetechie** [reply]
Gotcha. Not really sure I have anything to add on the SAF discussion, but I'm not quite understanding the case where you would ever have "thousands of disk files." But yes, if this is your use case, I can see how having a single file would be desirable. As @dnesbitt61 said though, only the tiles that are actually needed. In the hypothetical MD, PA, DE route, the majority of the trip will be on highways in the 0 level, which are 4x4 degrees in size. You'd only need to consult only a few tiles for the arterial and local roads. So the number of file ops is probably lower than what you imagine (unless you're scanning the directory of all files or something, which the actor might do at startup? I forget...).

Would be curious to see some profiling or other info if you're actually experiencing a problem with this. The main reason I can think of for MBTiles would be that it's easier to WRITE to if you want to expand your set of local tiles, but it does so at some extra complexity that's not really worth it unless it's a provable perf problem.

The main reason the tarball exists FWIW isn't so much to reduce the number of file handles needed (though that doesn't hurt) but rather to facilitate memory mapping a file which can be directly indexed into (so no compression IIRC? I may be wrong). This is a nice boon for server-side performance and startup times when you're serving continent or planet scale routing. But it doesn't apply to mobile. MBTiles or something like it does seem better suited to your use case if it's really a problem, but I'd suggest looking at ways to make it a mutable dataset instead.

**Reply by farfromrefug** [reply]
@dnesbitt61 indeed you are right only a few tiles are actually read per route computation. I was meaning having non tar file (thus thousand of files on disk). But indeed only a few would be read each time which would make SAf quite faster.
It would be indeed nice to have some profiling comparison. we should be able to create some by using  @Archdoog project and compare the time with CartoMobile SDK version.
And i am not saying MBTiles is the way to go. Just saying it is a nice way of handling it. Their might be other good ones. 
Mutable dataset might be a good thing yes. Not sure how easily it would be usable in mobile apps.
If you think as a user i think the best way to see offline rooting on mobile is to have "region" packages (whatever the format behind) , exactly like you do with offline maps in osmand and other map apps. This allows small package loading, small disk usage, and easy update mechanism. Not sure how you would do that with mutable dataset. I think it is the reason why Carto used MBTiles in the first place, it is the same workflow as for maps data. They actually also did the same for geocoding offline data

**Comment by HarelM** [comment]
I know this discussion is a bit old, but all the important people are here :-)
I'm helping @Archdoog and @ianthetechie to maintain valhalla-mobile.
I'm in the final process of integrating it into my app (Mapeak.com) for offline usage, which will allow downloading sections and also route between different downloaded adjacent areas.
If you are interested you can see the integration PR here:
https://github.com/IsraelHikingMap/Site/pull/2641

Since I plan to roll this to production soon I think the Rallista's repos are now production ready.

I was wondering if they can be on-boarded to the Valhalla org?
I still plan to continue maintain them, but I think it would be great to have an official home for them here.
This thread is a bit more about taking some things into valhalla core code, but my proposal is a lot "simpler" - just move the repos under the valhalla org.
The integration is fairly simple now on mobile devices thanks to @ianthetechie and @Archdoog, you add the relevant packages, use the relevant classes and vlahalla just works, which is great.
I've updated the docs, and handled a few PRs to improve coverage of valhalla's APIs.

Would be great to know what you think, especially @kevinkreiser and @nilsnolde.

**Reply by nilsnolde** [reply]
very cool, thanks!

> I still plan to continue maintain them, but I think it would be great to have an official home for them here.

who is "they"? I assume some repos:) we're very open to include (ideally fairly generically usable) valhalla libs/sdks/bindings etc. we're here on agreement on all sides, that for visibility it's best to pool the projects if possible (and wanted, sometimes I like having a repo under my own org/name:)).

I don't fully understand yet which repo(s) we're talking about exactly though

**Reply by HarelM** [reply]
Thanks for the prompt response.
The repos @Archdoog mentioned in his post - 4 repos total.
> The core project https://github.com/Rallista/valhalla-mobile/ which is published as a Swift Package and an android library on Maven Central. See the [README.md](https://github.com/Rallista/valhalla-mobile/blob/main/README.md) for a good overview.
https://github.com/Rallista/valhalla-openapi-models-swift - This provides all the valhalla API and Config JSON models (by no means 100% correct/finished). Giving the swift Valhalla-mobile library typed inputs and outputs from libvalhalla through the wrapper.
https://github.com/Rallista/valhalla-openapi-models-kotlin - Same as the swift package, but for kotlin.
https://github.com/stadiamaps/osrm-openapi - Same as the valhalla packages, but for OSRM flavored since most navigation SDKs lean that way.

**Reply by HarelM** [reply]
To be fair, I don't have maintainer access to stadiamaps, so just the 3 Rallista repos.

**Reply by nilsnolde** [reply]
ok cool! not sure about the osrm openapi one, guess that'd be better part of the osrm org:)

@kevinkreiser @chrstnbwnkl @kinkard wdyt?

**Reply by kinkard** [reply]
@nilsnolde last time we discussed this topic you've convinced me that all bindings should live inside Valhalla repo and not aside because of `git submodule` issues and simpler maintenance. So I would prefer merging https://github.com/Rallista/valhalla-mobile/ into Valhalla repo directly rather than keeping it in org and taking the responsibility to update Valhalla git submodule regularly.

Alternatively, if we decide that Swift/Kotlin bindings should live in a separate repo, I would initiate moving Python and NodeJS parts into separate projects as well.

**Reply by nilsnolde** [reply]
hm, yeah, you're right. I never had the time recently to look at it properly. if it's _only_ builds/bindings for mobile to valhalla, then perfect, we can indeed integrate it here! I somehow figured it'd do more, but that's then ferrostar I think. sorry, I promise to catch up on the whole mobile OSS space once I can freely breathe again :D

**Reply by nilsnolde** [reply]
maybe the openapi spec things can live in a single repo in the future in the valhalla org? they might be different, since it's different apis, but having them in a single repo still makes IMO. it's not build-heavy or so. then I'd rip them out here and move them there too.

**Reply by HarelM** [reply]
The openapi spec is probably the most obvious one. Yeah.
Whatever works for you as long as it's easy to consume. 😀
